src/echo/echo.rs

```diff
@@ -36,42 +36,23 @@ static LONG_HELP: &'static str = r#"
  \\xHH    byte with hexadecimal value HH (1 to 2 digits)
 "#;
 
-#[derive(Clone)]
-struct EchoOptions {
-    newline: bool,
-    escape: bool
-}
-
-#[inline(always)]
-fn to_char(bytes: &[u8], base: u32) -> char {
-    usize::from_str_radix(from_utf8(bytes.as_ref()).unwrap(), base).unwrap() as u8 as char
-}
-
-#[inline(always)]
-fn isxdigit(c: u8) -> bool {
-    match c as char {
-        '0' | '1' | '2' | '3' | '4' | '5' | '6' | '7' |
-        '8' | '9' | 'A' | 'B' | 'C' | 'D' | 'E' | 'F' => true,
-        _ => false
-    }
+enum Base {
+    B8 = 8,
+    B16 = 16,
 }
 
-#[inline(always)]
-fn isodigit(c: u8) -> bool {
-    match c as char {
-        '0' | '1' | '2' | '3' | '4' | '5' | '6' | '7' => true,
-        _ => false
-    }
+struct Opts {
+    newline: bool,
+    escape: bool
 }
 
-fn convert_str(string: &[u8], index: usize, base: u32) -> (char, usize) {
+fn convert_str(string: &[u8], index: usize, base: Base) -> (char, usize) {
     let (max_digits, is_legal_digit) : (usize, fn(u8) -> bool) = match base {
-        8 => (3, isodigit),
-        16 => (2, isxdigit),
-        _ => panic!(),
+        Base::B8 => (3, |c| { (c as char).is_digit(8) }),
+        Base::B16 => (2, |c| { (c as char).is_digit(16) }),
     };
 
-    let mut bytes = vec!();
+    let mut bytes = vec![];
     for offset in 0usize .. max_digits {
```

Comment:
Remove the spaces around the `..` (and fix elsewhere in the file).  I also assume you can change the `0usize` to `0`.